packages/peregrine/lib/talons/Cms/useCmsPage.js

```diff
@@ -0,0 +1,70 @@
+import { useEffect, useMemo } from 'react';
+import { useQuery } from '@apollo/react-hooks';
+import { useAppContext } from '../../context/app';
+
+/**
+ * Retrieves data necessary to render a CMS Page
+ *
+ * @param {object} props
+ * @param {object} props.id - CMS Page ID
+ * @param {object} props.queries - Collection of GraphQL queries
+ * @param {object} props.queries.getCmsPage - Query for getting a CMS Page
+ * @returns {{shouldShowLoadingIndicator: *, hasContent: *, cmsPage: *, error: *}}
+ */
+export const useCmsPage = props => {
+    const {
+        id,
+        queries: { getCmsPage }
+    } = props;
+
+    const { loading, error, data } = useQuery(getCmsPage, {
+        variables: {
+            id: Number(id)
+        },
+        fetchPolicy: 'cache-and-network'
+    });
+
+    const [
+        { isPageLoading },
+        {
+            actions: { setPageLoading }
+        }
+    ] = useAppContext();
+
+    // To prevent loading indicator from getting stuck, unset on unmount.
+    useEffect(() => {
+        return () => {
+            setPageLoading(false);
+        };
+    }, [setPageLoading]);
+
+    // Ensure we mark the page as loading while we check the network for updates
+    useEffect(() => {
+        if (loading && !isPageLoading) {
+            setPageLoading(true);
+        } else if (!loading && isPageLoading) {
+            setPageLoading(false);
+        }
```

Comment:
This pattern works, but it's also overkill. If `setPageLoading` is called with the same value that's already in state, it doesn't count as a state change (i.e., doesn't cause an update and re-render). So these five lines can be one, and the dependencies don't need to include `isPageLoading`.

```es6
useEffect(
  () => { setPageLoading(loading) },
  [loading, setPageLoading]
)
```

Reply:
That's smart, thanks for explaining that!